### PR TITLE
Fix multiplication and division test failures in 330db and 340 shim [databricks]

### PR DESCRIPTION
**Pull request by razajafri**
This PR addresses division and multiplication test failures in 11.3 databricks and Spark 340

Before this PR, running arithmetic_ops_tests we had
`338 failed, 464 passed, 14 skipped, 2 xfailed in 448.22s (0:07:28)`

After 
`214 failed, 588 passed, 14 skipped, 2 xfailed in 651.23s (0:10:51)`

Contributes towards https://github.com/NVIDIA/spark-rapids/issues/7348

### CODE REVIEW

**Review by gerashegalov — CHANGES_REQUESTED** [review]
verify-all-modules-with-headSparkVersion PR check failed
> found modified files during mvn verify:
 M docs/supported_ops.md
 M tools/generated_files/supportedExprs.csv

generate markdown and update the PR

File: sql-plugin/src/main/311until340-non330db/scala/org/apache/spark/sql/rapids/arithmetic.scala [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  override def hasSideEffects: Boolean = super.hasSideEffects || SQLConf.get.ansiEnabled
+}
+
+case class GpuIntegralDivide(
+    left: Expression,
+    right: Expression) extends GpuIntegralDivideParent(left, right)
+
+case class GpuRemainder(left: Expression, right: Expression) extends GpuRemainderParent(left, right)
+
+case class GpuPmod(left: Expression, right: Expression) extends GpuPmodParent(left, right)
```
Comment (razajafri) [inline sql-plugin/src/main/311until340-non330db/scala/org/apache/spark/sql/rapids/arithmetic.scala:38]: This and other references of `*Parent` will be fixed to Base after #7355 is merged

**Review by gerashegalov — DISMISSED** [review]
LGTM

**Comment by razajafri** [comment]
build

**Comment by gerashegalov** [comment]
@razajafri a tiered projection test needs to be addressed
```
[2022-12-16T05:11:18.570Z] - Get Expression Tiers - Query derived from nds q4 *** FAILED ***
```

**Comment by razajafri** [comment]
> @razajafri a tiered projection test needs to be addressed
> 
> ```
> [2022-12-16T05:11:18.570Z] - Get Expression Tiers - Query derived from nds q4 *** FAILED ***
> ```

I have pushed a fix for this. Please correct me if I am wrong, but I think this test wasn't written correctly and the reason it was passing before was because GpuDecimalMultiply wasn't extending CudfBinaryArithmetic, and had `checkInputDataTypes()` from `Expression` which returned true without testing the inputs. 

`def checkInputDataTypes(): TypeCheckResult = TypeCheckResult.TypeCheckSuccess`

Here is what I see in spark-shell (Spark-3.1.1)

```
scala> val df = Seq(("1234567.21", "1234154"), ("9223372.08", "1")).toDF.withColumn("dec1", col("_1").cast(DecimalType(7,2))).withColumn("dec2", col("_2").cast(DecimalType(10,0)))
df: org.apache.spark.sql.DataFrame = [_1: string, _2: string ... 2 more fields]

scala> df.show
+----------+-------+----+-------+
|        _1|     _2|dec1|   dec2|
+----------+-------+----+-------+
|1234567.21|1234154|null|1234154|
|9223372.08|      1|null|      1|
+----------+-------+----+-------+


scala> df.printSchema
root
 |-- _1: string (nullable = true)
 |-- _2: string (nullable = true)
 |-- dec1: decimal(7,2) (nullable = true)
 |-- dec2: decimal(10,0) (nullable = true)

scala> df.selectExpr("dec1 * dec2").printSchema
root
 |-- (dec1 * dec2): decimal(18,2) (nullable = true)


scala> df.selectExpr("dec1 * dec2").explain(true)
== Parsed Logical Plan ==
'Project [('dec1 * 'dec2) AS (dec1 * dec2)#72]
+- Project [_1#37, _2#38, dec1#44, cast(_2#38 as decimal(10,0)) AS dec2#48]
   +- Project [_1#37, _2#38, cast(_1#37 as decimal(7,2)) AS dec1#44]
      +- LocalRelation [_1#37, _2#38]

== Analyzed Logical Plan ==
(dec1 * dec2): decimal(18,2)
Project [CheckOverflow((promote_precision(cast(dec1#44 as decimal(12,2))) * promote_precision(cast(dec2#48 as decimal(12,2)))), DecimalType(18,2), true) AS (dec1 * dec2)#72]
+- Project [_1#37, _2#38, dec1#44, cast(_2#38 as decimal(10,0)) AS dec2#48]
   +- Project [_1#37, _2#38, cast(_1#37 as decimal(7,2)) AS dec1#44]
      +- LocalRelation [_1#37, _2#38]

== Optimized Logical Plan ==
LocalRelation [(dec1 * dec2)#72]

== Physical Plan ==
LocalTableScan [(dec1 * dec2)#72]
```

As we can see, Spark wraps the inputs in checkoverflow and promoteprecision. But if we do that, the test will start failing for Spark 340. So what I have done is simply casted the input types to how Spark expects them to be.

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
Looking into the CI failure, it's related to the last failure

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
I forgot to split the GpuAverage work to it's own shim. 

Will post an update soon

**Comment by razajafri** [comment]
build

**Review by gerashegalov — APPROVED** [review]
LGTM

**Pull request merged**